core/src/price_reporter/exchange/uni_v3.rs

```diff
@@ -98,171 +84,171 @@ impl UniswapV3Handler {
-            if a.block_number < b.block_number {
-                return Ordering::Less;
-            } else if a.block_number > b.block_number || a.transaction_index < b.transaction_index {
-                return Ordering::Greater;
-            } else if a.transaction_index > b.transaction_index {
-                return Ordering::Less;
-            }
-            Ordering::Equal
-        });
-        let swap = swap_filter_recent_events.pop();
-        if let Some(swap) = swap {
-            let block_id = BlockId::Number(BlockNumber::Number(swap.block_number.unwrap()));
-            let block_timestamp = web3_connection
-                .eth()
-                .block(block_id)
-                .await
-                .unwrap()
-                .unwrap()
-                .timestamp;
-            let price_report = Self::handle_event(
-                base_token.clone(),
-                quote_token.clone(),
-                swap,
-                is_flipped,
-                swap_event_abi.clone(),
-            );
-            if let Some(mut price_report) = price_report {
-                price_report.local_timestamp = get_current_time();
-                price_report.reported_timestamp = Some(block_timestamp.as_u128());
-                sender.send(price_report).unwrap();
-            }
         }
+    };
+}
 
-        // Start streaming events from the swap_filter.
-        let base_token_clone = base_token.clone();
-        let quote_token_clone = quote_token.clone();
-        let worker_handle = tokio::spawn(async move {
-            let swap_stream = swap_filter.stream(Duration::new(1, 0));
-            futures::pin_mut!(swap_stream);
-            loop {
-                let swap = swap_stream.next().await.unwrap().unwrap();
-                let block_id = BlockId::Number(BlockNumber::Number(swap.block_number.unwrap()));
-                let block_timestamp = web3_connection
-                    .eth()
-                    .block(block_id)
-                    .await
-                    .unwrap()
-                    .map(|block| block.timestamp.as_u128());
-                let price_report = Self::handle_event(
-                    base_token_clone.clone(),
-                    quote_token_clone.clone(),
-                    swap,
-                    is_flipped,
-                    swap_event_abi.clone(),
-                );
-                if let Some(mut price_report) = price_report {
-                    price_report.local_timestamp = get_current_time();
-                    price_report.reported_timestamp = block_timestamp;
-                    sender.send(price_report).unwrap();
-                }
-            }
-        });
+/// The core handler for UniswapV3, responsible for defining Swap event filters, streaming events,
+/// and parsing as PriceReports.
+pub struct UniswapV3Connection {
+    /// The underlying price stream
+    price_stream: Box<dyn Stream<Item = Price> + Unpin>,
```

Comment:
We still have to implement `poll_next` because we want these structs themselves to implement `Stream`. I'll change it to `InitializablePriceStream` though to avoid dynamic lookup